### PR TITLE
login: T4754: show configured 2FA OTP key

### PR DESCRIPTION
**Pull request by goodNETnick**
## Change Summary

Added ability to display the configured user's OTP settings in the VyOS CLI.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4754

## Component(s) name

system, login

## Proposed changes

Operational mode command added:
`show system login authentication user <username> otp <full | key-b32 | qrcode | uri >`

## How to test

```
vyos@vyos:~$ show system login authentication user tst otp full
# You can share it with the user, he just needs to scan the QR in his OTP app
# username:  tst
# OTP KEY:  5PA4SGYTQSGOBO3H3EQSSNCUNZAYAPH2
# OTP URL:  otpauth://totp/tst@vyos?secret=5PA4SGYTQSGOBO3H3EQSSNCUNZAYAPH2&digits=6&period=30
█████████████████████████████████████████
█████████████████████████████████████████
████ ▄▄▄▄▄ █▀ ██▄▀ ▄█▄▀▀▄▄▄▄██ ▄▄▄▄▄ ████
████ █   █ █▀ █▄▄▀▀▀▄█  ▄▄▀▄ █ █   █ ████
████ █▄▄▄█ █▀█▀▄▄▀  ▄▀ █▀ ▀▄██ █▄▄▄█ ████
████▄▄▄▄▄▄▄█▄█▄▀ ▀▄█ ▀ ▀ ▀ █▄█▄▄▄▄▄▄▄████
████  ▄▄▄▀▄▄  ▄███▀▄▀█▄██▀ ▀▄ ▀▄█ ▀ ▀████
████ ▀▀ ▀ ▄█▄ ▀ ▀▄ ▄█▀ ▄█ ▄▀▀▄██    █████
████▄ █▄▀▀▄█▀ ▀█▄█▄▄▄▄ ▄▀█▀▀█ ▀ ▄ ▀█▀████
█████  ▀█▀▄▄ █ ▀▄▄  ▄█▄    ▀█▀▀ █▀ ▄█████
████▀██▀█▄▄ ▀▀▀▀█▄▀ ▀█▄▄▀▀▀ ▀ ▀█▄██▀▀████
████▄ ▄ ▄▀▄██▀█ ▄ ▀▄██ ▄▄  ▀▀▄█▄██ ▄█████
████▀▀ ▄▀ ▄ ▀█▀█▀█  █▀█▄▄▀█▀█▄██▄▄█ ▀████
████ █ ▀█▄▄█▄ ▀ ▄▄▀▀  ▀ █▄█▀████ █▀ ▀████
████▄██▄██▄█▀ ▄▀ ▄▄▀▄  ▄▀█ ▄ ▄▄▄ ▀█▄ ████
████ ▄▄▄▄▄ █▄  ▀█▄█ ▄ ▀ ▄ ▄  █▄█ ▄▀▄█████
████ █   █ █ ▀▄██▄▄▀█▄▀▄██▄▀  ▄  ▀██▀████
████ █▄▄▄█ █ ██▀▄▄  ▀▄▄▀█▀ ▀█ ▄▀█ ▀██████
████▄▄▄▄▄▄▄█▄███▄███▄█▄▄▄▄█▄▄█▄██▄█▄█████
█████████████████████████████████████████
█████████████████████████████████████████
# To add this OTP key to configuration, run the following commands:
set system login user otptester authentication otp key '5PA4SGYTQSGOBO3H3EQSSNCUNZAYAPH2'
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

File: op-mode-definitions/show-system.xml.in [hunk 1]
```diff
@@ -66,6 +66,55 @@
...
+              <node name="authentication">
+                <properties>
+                  <help>Show user account authentication information</help>
+                </properties>
+                <children>
+                  <tagNode name="user">
+                    <properties>
+                      <help>Show configured user</help>
+                      <completionHelp>
+                        <script>${vyos_libexec_dir}/vyos-sudo.py ${vyos_op_scripts_dir}/show_users.py | awk '{print $1}' | tail -n +3</script>
```
Comment (sever-sever) [inline op-mode-definitions/show-system.xml.in]: Can we use `path` ?
```
<path>system login user</path>
```

**Review by sever-sever — CHANGES_REQUESTED** [review]

**Comment by sever-sever** [comment]
All new op-mode scripts should be written in the new `vyos.opmode` format
https://phabricator.vyos.net/T4564
https://phabricator.vyos.net/T2719

Is it possible to use the new format for op-mode?

File: op-mode-definitions/show-system.xml.in (hunk 1 above)
Comment (goodNETnick) [reply op-mode-definitions/show-system.xml.in]: Thank you!
It's really much better that way :)
Fixed

**Comment by c-po** [comment]
IMHO this should actually be a generate command, as it generates a new OTP token.

**Comment by goodNETnick** [comment]
Despite the fact that what the user sees in the end is very visually similar,
generate and show are completely different tasks:
- generate a new key with commands
- show an existing key, for example to show a barcode
Generator:
https://github.com/vyos/vyos-1x/pull/1599

**Comment by c-po** [comment]
> Despite the fact that what the user sees in the end is very visually similar, generate and show are completely different tasks:
> 
> * generate a new key with commands
> * show an existing key, for example to show a barcode
>   Generator:
>   [login: T4751: 2FA OTP key generator in VyOS CLI #1599](https://github.com/vyos/vyos-1x/pull/1599)

Ah, please combine both PRs so this does not confuse any more people. Please also use the new `vyos.opmode` format as mentioned by @sever-sever.

**Comment by goodNETnick** [comment]
I'd rather not make one big PR, but split it into 2
The tasks and solutions are completely different
Regarding the new op-mode, I will modify the solution

**Comment by c-po** [comment]
> I'd rather not make one big PR, but split it into 2

I think you mix up a PR and a commit. A PullRequest can contain an unlimited amount of commits. Indeed you shout use two commits (one for show and one for generate)

**Comment by goodNETnick** [comment]
> All new op-mode scripts should be written in the new `vyos.opmode` format https://phabricator.vyos.net/T4564 https://phabricator.vyos.net/T2719
> 
> Is it possible to use the new format for op-mode?

Rewritten to new format

**Comment by sever-sever** [comment]
I guess we can merge it
And then fix all that we need.

**Comment by sempervictus** [comment]
+1 to accelerate wide adoption of integrated MFA - the "dedicated providers" are being breached by children, messed up by M&A, and so on... having this function encapsulated an in a controlled environment seems pressing.

**Comment by sever-sever** [comment]
> +1 to accelerate wide adoption of integrated MFA - the "dedicated providers" are being breached by children, messed up by M&A, and so on... having this function encapsulated an in a controlled environment seems pressing.

@sempervictus  We are going to revert MFA OTP as it brokes RADIUS authentication https://vyos.dev/T4943
Didn't find solution for it

**Review by S-code652 — APPROVED** [review]

**Review by S-code652 — APPROVED** [review]

**Comment by dmbaturin** [comment]
Let's merge it when 1.4 is branched off.

**Comment by sever-sever** [comment]
Merged: if something is wrong, we'll fix it.

**Comment by sempervictus** [comment]
@sever-sever - anything you want us humans to test from this merge?

**Comment by sever-sever** [comment]
> @sever-sever - anything you want us humans to test from this merge?

@sempervictus  Nothing specific :)
Basic functions to show qr and other
Thanks

**Pull request merged**